<a href="https://colab.research.google.com/github/jazaineam1/Andes_ADDBD/blob/main/Sesion4/M1_Clase_4_Manipulaci%C3%B3n_de_datos_en_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ***Análisis de Datos y Big data***

## ***Universidad de los Andes***
>## **Dirección de Educación Continua.**
>>## **Formación de habilidades escenciales.**

![Imágen1](https://github.com/jazaineam1/Andes_ADDBD/blob/main/Imagenes/Presentaci%C3%B3n-removebg-preview.png?raw=true)

## **Manipulación de datos en Python**
## **Módulo 1**

## ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?
En el desarrollo de este caso introduciremos las librerías pandas, numpy y matplotlib para el preprocesamiento, visualización de un conjunto de datos y el desarrollo de algunas medidas estadísticas descriptivas.



##  Introducción

La revisión del comportamiento de los clientes en cualquier empresa es de vital importancia para el desarrollo de estrategias  tanto de mejoramiento de productos como de publicidad y toma de decisiones. Evaluaremos la información obtenida de un conjunto de registro obtenidos de una empresa de telecomunicaciones que cuenta con la revisión de las siguientes variables:

1. totrev: Ingresos del cliente
2. totmou: Total de minutos usados por el cliente
3. area: Área geográfica
4. creditcd: Indicador de tarjeta de crédito
5. eqpdays: Número de días (antigüedad) del equipo actual

Es importante mencionar que la empresa contiene una gran cantidad de clientes y hacer el análisis uno a uno no es lo  óptimo. Con ayuda de la programación y el análisis de datos, se logrará llegar a conclusiones rápidas y acertadas para encontrar relaciones que a simple vista no son tenidas en cuenta.

### ¿Qué se quiere identificar? Teniendo en cuenta la información de la base de datos se quiere identificar si hay una relación entre el total de minutos usados por el cliente con la compañía y el ingreso total de los clientes, es decir, Será posible afirmar que cuando una persona
tiene un mayor ingreso mayor entonces es mayor la utilización de minutos o todo lo contrario?


####  Insumos para tratar este problema


Teniendo en cuenta que la compañía tenia estándares para la entrega de la información, se dispone de un  conjunto de tablas en formato CSV ordenado de la siguiente manera:
La información de este conjunto de datos puede encontrase en [aquí](https://www.kaggle.com/abhinav89/telecom-customer?select=Telecom_customer+churn.csv)

### Objetivos

En este caso usted tendrá que cargar varias bases de datos, hará una exploración básica sobre la información y fusionará las distintas bases para tener una visión general del problema. La idea es que desarrolle las siguientes habilidades:

1. Manejo de la librería pandas de Python para cargar y leer datos;
2. Ideas elementales para una útil transformación de los datos;
3. Construcción y presentación de argumentos válidos que le brindarán una solución para la pregunta expresada más arriba y llegar a una conclusión.
4. Visualización de gráficas elementales que le permitirá entender mejor la información contenida en los datos.

#### Importación de paquetes


Una de las mejores opciones para trabajar con datos tabulares en Python es usar el módulo pandas. La librería `pandas` provee estructuras de datos, genera gráficos de alta calidad con `matplotlib` y se integra de buena forma con otras librerías que usan arrays de `numpy`.

Debemos revisar si la librería se encuentra en nuestro sistema usando el comando

```python
!pip show librería
```

In [ ]:
!pip show pandas
!pip show numpy
!pip show matplotlib

Name: pandas
Version: 1.1.1
Summary: Powerful data structures for data analysis, time series, and statistics
Home-page: https://pandas.pydata.org
Author: None
Author-email: None
License: BSD
Location: c:\users\antonino\anaconda3\ana\lib\site-packages
Requires: numpy, python-dateutil, pytz
Required-by: statsmodels, seaborn, plotly-express, Orange3, geopandas, dplython, dfply, bqplot, AeroSandbox
Name: numpy
Version: 1.19.1
Summary: NumPy is the fundamental package for array computing with Python.
Home-page: https://www.numpy.org
Author: Travis E. Oliphant et al.
Author-email: None
License: BSD
Location: c:\users\antonino\anaconda3\ana\lib\site-packages
Requires: 
Required-by: tables, statsmodels, seaborn, scipy, scikit-learn, PyWavelets, python-louvain, pyqtgraph, plotly-express, Pattern, patsy, pandas, Orange3, openTSNE, numexpr, numba, mkl-random, mkl-fft, matplotlib, imageio, h5py, dplython, dfply, bqplot, Bottleneck, bokeh, bkcharts, astropy, AeroSandbox
Name: matplotlib
Version: 3.

En el caso de que la librería no se encuentre instalada ejecutamos el comando

```python
!pip install librería
```

Para importar las librerías usamos el comando
```python
import librería as alias
```

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

Luego de que tener los paquetes activados dentro de nuestro ambiente de python podremos usar las fucnciones que se encuentren en la librería.

### Introducción a pandas

Poner información sobre pandas e indicar la diferencia entre series y dataframe

![](https://www.cdn.geeksforgeeks.org/wp-content/uploads/creating_dataframe1.png)

A continuación, mostramos el código a ejecutar para crear una ` Serie`

In [ ]:
Serie1 = pd.Series(
index=[0, 1, 2, 3, 4], name="Variable1", data=["A","B","C","D","E"]
)
Serie1

0    A
1    B
2    C
3    D
4    E
Name: Variable1, dtype: object

In [ ]:
Data_Frame = pd.DataFrame(
index=[0, 1, 2, 3, 4], columns=["Variable2","Variable3"], data=[[1, 1], ["2", 2],[3,3],[4,4],[5,5]]
)
Data_Frame

Variable2  Variable3
0         1          1
1         2          2
2         3          3
3         4          4
4         5          5

Para revisar el tipo de objeto que contiene nuestro DataFrame podemos ejecutar

In [ ]:
Data_Frame.dtypes
#Serie1.dtypes

Variable2    object
Variable3     int64
dtype: object

### Carga de información en `pandas`

Para nuestro  ejercicio usaremos los siguientes conjuntos de datos alojado [acá](https://github.com/jazaineam1/Andes_ADDBD/blob/main/Sesion4/Bases.zip), por favor descarguelo y deje sus datos en el archivo colab:

1. NEW ENGLAND AREA.csv
2. ATLANTIC SOUTH AREA.csv
3. NEW YORK CITY AREA.csv
4. CHICAGO AREA.csv
5. NORTH FLORIDA AREA.csv
6. DC-MARYLAND-VIRGINIA AREA.csv
7. GREAT LAKES AREA.csv
8. NORTHWEST-ROCKY MOUNTAIN AREA.csv
9. MIDWEST AREA.csv
10. SOUTHWEST AREA.csv
11. LOS ANGELES AREA.csv
12. HOUSTON AREA.csv
13. CALIFORNIA NORTH AREA.csv
14. CENTRAL-SOUTH TEXAS AREA.csv
15. DALLAS AREA.csv
16. PHILADELPHIA AREA.csv
17. TENNESSEE AREA.csv
18. OHIO AREA.csv
19. SOUTH FLORIDA AREA.csv

Cada uno de estos conjuntos de datos contiene la información de:

1. *mou_Mean:* Número medio de minutos de uso mensuales
2. *custcare_Mean:* Número medio de llamadas de atención al cliente
3. *area:*  Área geografica
4. *creditcd:* Indicador de tarjeta de crédito
5. *eqpdays:* Número de días (antigüedad) del equipo actual

Además, se incluye _Credito_dias.csv_ con la información de 2 variables adicionales para los registros las cuales son:

1. _creditcd:_ Indicador de tarjeta de crédito
2. _eqpdays:_ Número de días (antigüedad) del equipo actual

Revisaremos inicialmente los datos contenidos en _Credito_dias.csv_ para revisar algunas funciones interesantes de `pandas`, usaremos el comando
```python
pd.read_csv("archivo.csv")
```
Podemos definir el número de registros a imprimir usando
```python
pd.options.display.min_rows=10
````

In [34]:
df=pd.read_csv("Bases/Credito_dias.csv", sep=",",index_col="Customer_ID")
df.head(7)

creditcd  eqpdays
Customer_ID                  
1021919            N      127
1061499            N      471
1089004            Y      289
1014402            Y      867
1014503            N      269
1010267            Y      163
1074203            N      238

In [ ]:
help(pd.read_csv)

El conjunto de datos se encuentra en el `DataFrame` `df`

sobre el objeto creado podemos usar  algunas funciones  para darnos una idea del comportamiento de la información:
```python
df.head(n)  # imprime los primeros n registros del DataFrame
df.tail(n)  # imprime los primeros n registros del DataFrame
df.shape    # imprime el numero de columnas y filas del DataFrame
df.columns  # imprime el nombre de las columnas del DataFrame
df.index    # imprime el indice de los registros del DataFrame
df.dtypes   # inpreme el tipo de cada una de las columnnas del DataFrame
df.sample(n)# imprime una muestra aleatoria de n registros en el DataFrame
```

In [37]:
display("primeros 3 registros",df.head(3))
display("últimos registros",df.tail(3))
display(df.columns)
display(df.index)
display("Tipos de registros en el DataFrame",df.dtypes)
display("Muestra akleatoria de 3 registros",df.sample(3))
display("dimensión",df.shape)

'primeros 3 registros'

creditcd  eqpdays
Customer_ID                  
1021919            N      127
1061499            N      471
1089004            Y      289

'últimos registros'

creditcd  eqpdays
Customer_ID                  
1006191            Y      953
1076628            N      285
1038465            N      472

Index(['creditcd', 'eqpdays'], dtype='object')

Index([1021919, 1061499, 1089004, 1014402, 1014503, 1010267, 1074203, 1073093,
       1005754, 1090260,
       ...
       1050240, 1069471, 1002645, 1088508, 1081270, 1094996, 1065953, 1006191,
       1076628, 1038465],
      dtype='int64', name='Customer_ID', length=10000)

'Tipos de registros en el DataFrame'

creditcd    object
eqpdays      int64
dtype: object

'Muestra akleatoria de 3 registros'

creditcd  eqpdays
Customer_ID                  
1095397            N      186
1062351            N      296
1075162            N      176

'dimensión'

(10000, 2)

In [60]:
df['totmou']=np.random.randint(1,100,len(df))
df['totrev']=np.random.randint(1,100,len(df))

In [76]:
df[df["totrev"]<5]

creditcd  eqpdays  totmou  totrev
Customer_ID                                  
1074203            N      238       5       3
1091755            N      199      71       3
1094835            N      259      66       1
1085151            Y      362      94       1
1046268            N      488      40       3
...              ...      ...     ...     ...
1046377            N      425      93       4
1059143            Y      493      21       3
1028305            Y      643      14       1
1039843            Y       87      66       2
1030796            Y      632      46       1

[376 rows x 4 columns]

Este conjunto de datos corresponde a 10000 registros y 2 variables

### Manejo de datos duplicados

os dataframes tienen la posibilidad de detectar las filas duplicadas con la función:
```
df.duplicated({columns})
```
Si no especifica {columns}, se busacarán duplicados teniendo en cuenta todas las columnas. Una vez detectados el dtaframe tiene un función para eliminar filas duplicadas, haciendo:

```
df = df.drop_duplicates()
```
También es posible eliminar filas que duplican solo algunos campos. Con el fin de dejar solo una ocurrencia.

```
df.drop_duplicates(['nombre', 'apellido', 'cédula'], keep='last')
```
En este caso, como no todas las colomnas son iguales se conserva solo la última ocurrencia.

### Manejo de datos perdidos o esperados
Son aquellos datos flatantes en la tabla que por alguna razón o error no se encuentran, o parecen con algún valor o etiqueta de no válido.
Para manipular datos perdidos use la explicación dada el siguiente enlace: [DATOS PERDIDOS](http://ligdigonzalez.com/manipulando-datos-perdidos-en-python/)
recuerde que para reemplazar datos perdidos en un dataframe $df$ se usa la función:
```python
promedio = df['totrev'].mean()
df['nota'].replace(np.nan, promedio)
```
en donde la opción 'promedio' busca el valor promedio de la columna (datos numéricos) o la moda (datos categóricos) para reemplazar el valor no válido.

In [77]:
df.isnull().any(axis=0)

creditcd     True
eqpdays     False
totmou      False
totrev      False
dtype: bool

In [39]:
df.isnull().any(axis=1)

Customer_ID
1021919    False
1061499    False
1089004    False
1014402    False
1014503    False
           ...  
1094996    False
1065953    False
1006191    False
1076628    False
1038465    False
Length: 10000, dtype: bool

In [87]:
df["creditcd"].isnull()

Customer_ID
1021919    False
1061499    False
1089004    False
1014402    False
1014503    False
           ...  
1094996    False
1065953    False
1006191    False
1076628    False
1038465    False
Name: creditcd, Length: 10000, dtype: bool

In [91]:
perdidos = df[(df['creditcd'].isnull()) ]
perdidos

creditcd  eqpdays  totmou  totrev
Customer_ID                                  
1031546          NaN      736      95      46
1065654          NaN      351      73      78
1080786          NaN      312      93      73
1088854          NaN      227      80      90
1025529          NaN      670      33      19
...              ...      ...     ...     ...
1059882          NaN      321      74      55
1096563          NaN      209       9      55
1074077          NaN      292      34      38
1073873          NaN       13      80      42
1092015          NaN      243      89      93

[181 rows x 4 columns]

In [80]:
promedio = df['totrev'].mean()
df['totrev'].replace(np.nan, promedio)

Customer_ID
1021919    50
1061499    55
1089004    85
1014402    73
1014503     7
           ..
1094996    50
1065953    68
1006191     7
1076628    65
1038465     8
Name: totrev, Length: 10000, dtype: int64

También es recomendado usar la funcion `fillna()`

In [92]:
df['creditcd'].fillna("Y",inplace=True)

In [93]:
df.isnull().any(axis=0)

creditcd    False
eqpdays     False
totmou      False
totrev      False
dtype: bool

In [94]:
df

creditcd  eqpdays  totmou  totrev
Customer_ID                                  
1021919            N      127      81      50
1061499            N      471      71      55
1089004            Y      289      76      85
1014402            Y      867      57      73
1014503            N      269      32       7
...              ...      ...     ...     ...
1094996            N      194      32      50
1065953            N      394      69      68
1006191            Y      953      50       7
1076628            N      285       3      65
1038465            N      472      55       8

[10000 rows x 4 columns]



### Cálculo de estadísticas resumen

Las medidas estadísticas resumen son de gran ayuda para comprender como es el comportamiento de la distribución de los datos, los objetos `DataFrame` ofrecen facilidades para hacer el cálculo de algunas estadísticas sobre cada una de las variables.

```python
df["var"].value_counts() # Encuentra la frecuencia de las categorias de una columna (la columna debe ser cualitativa)
df["var"].min()          # Encuentra el mínimo de una variable cuantitativa
df["var"].median()       # Encuentra el mínimo de una variable cuantitativa
df["var"].mean()         # Encuentra el promedio de una variable cuantitativa
df["var"].max()          # Encuentra el máximo de una variable cuantitativa
df["var"].quantile()     # Encuentra los cuantiles de una variable cuantitativa
```


In [95]:
df["creditcd"].value_counts()

creditcd
Y    6970
N    3030
Name: count, dtype: int64

In [96]:
df["creditcd"].count()

10000

In [97]:
df["eqpdays"].min()

-5

La exploración del anterior resultado puede mostrar inconsistencias  en los registros y nos ayudará a tomar decisiones sobre algunos registros, el valor mínimo de la variable `eqpdays` (Número de días (antigüedad) del equipo actual) debería ser estrictamente positivo y estos pueden ser errores de ingreso.

In [98]:
df["eqpdays"].max()

1812

In [99]:
df["eqpdays"].median()

340.0

In [100]:
df["eqpdays"].mean()

393.468

In [101]:
df["eqpdays"].quantile(0.25)

214.0

In [102]:
df["eqpdays"].quantile(0.75)

528.0

Estas estadísticas pueden ser resumidas con el comando:
```python
df["var"].describe()  # Estadísticas resumen principales
```

In [103]:
df["eqpdays"].describe()

count    10000.000000
mean       393.468000
std        257.180316
min         -5.000000
25%        214.000000
50%        340.000000
75%        528.000000
max       1812.000000
Name: eqpdays, dtype: float64

### Agregar información de múltiples tablas

Hemos desarrollado un breve resumen estadístico solamente usando la base `Credito_dias.csv`. Realizaremos la combinación de las 19 áreas que nos reportan en los archivos csv. Una forma de lograr esta tarea de agregación es usar el método pd.concat() de pandas. Una entrada en este método puede ser una lista de DataFrames que quiera concatenar. Usaremos un ciclo  `for` sobre cada uno de los archivos de áreas para


In [105]:
df.columns


Index(['creditcd', 'eqpdays', 'totmou', 'totrev'], dtype='object')

In [106]:
df.dtypes

creditcd    object
eqpdays      int64
totmou       int64
totrev       int64
dtype: object

### Ejercicio 1
Determinar los percentiles 25, 50 y 75 para las columnas ttomou, totrev, eqpdays

In [ ]:
#Espacio para el desarrollo del ejercicio

In [107]:
os.listdir("Bases")

['CHICAGO AREA.csv',
 'SOUTH FLORIDA AREA.csv',
 'SOUTHWEST AREA.csv',
 'HOUSTON AREA.csv',
 'CENTRAL-SOUTH TEXAS AREA.csv',
 'GREAT LAKES AREA.csv',
 'CALIFORNIA NORTH AREA.csv',
 'NORTHWEST-ROCKY MOUNTAIN AREA.csv',
 'NORTH FLORIDA AREA.csv',
 'OHIO AREA.csv',
 'MIDWEST AREA.csv',
 'PHILADELPHIA AREA.csv',
 'Credito_dias.csv',
 'DALLAS AREA.csv',
 'DC-MARYLAND-VIRGINIA AREA.csv',
 'TENNESSEE AREA.csv',
 'NEW ENGLAND AREA.csv',
 'NEW YORK CITY AREA.csv',
 'LOS ANGELES AREA.csv',
 'ATLANTIC SOUTH AREA.csv']

In [110]:
listado=os.listdir("Bases")
listado

['CHICAGO AREA.csv',
 'SOUTH FLORIDA AREA.csv',
 'SOUTHWEST AREA.csv',
 'HOUSTON AREA.csv',
 'CENTRAL-SOUTH TEXAS AREA.csv',
 'GREAT LAKES AREA.csv',
 'CALIFORNIA NORTH AREA.csv',
 'NORTHWEST-ROCKY MOUNTAIN AREA.csv',
 'NORTH FLORIDA AREA.csv',
 'OHIO AREA.csv',
 'MIDWEST AREA.csv',
 'PHILADELPHIA AREA.csv',
 'Credito_dias.csv',
 'DALLAS AREA.csv',
 'DC-MARYLAND-VIRGINIA AREA.csv',
 'TENNESSEE AREA.csv',
 'NEW ENGLAND AREA.csv',
 'NEW YORK CITY AREA.csv',
 'LOS ANGELES AREA.csv',
 'ATLANTIC SOUTH AREA.csv']

In [111]:
listado.remove('Credito_dias.csv')

In [112]:
listado

['CHICAGO AREA.csv',
 'SOUTH FLORIDA AREA.csv',
 'SOUTHWEST AREA.csv',
 'HOUSTON AREA.csv',
 'CENTRAL-SOUTH TEXAS AREA.csv',
 'GREAT LAKES AREA.csv',
 'CALIFORNIA NORTH AREA.csv',
 'NORTHWEST-ROCKY MOUNTAIN AREA.csv',
 'NORTH FLORIDA AREA.csv',
 'OHIO AREA.csv',
 'MIDWEST AREA.csv',
 'PHILADELPHIA AREA.csv',
 'DALLAS AREA.csv',
 'DC-MARYLAND-VIRGINIA AREA.csv',
 'TENNESSEE AREA.csv',
 'NEW ENGLAND AREA.csv',
 'NEW YORK CITY AREA.csv',
 'LOS ANGELES AREA.csv',
 'ATLANTIC SOUTH AREA.csv']

In [45]:
print("Definición de los símbolos de las acciones")
areas = ['ATLANTIC SOUTH AREA',
 'CALIFORNIA NORTH AREA',
 'CENTRAL-SOUTH TEXAS AREA',
 'CHICAGO AREA',
 'DALLAS AREA',
 'DC-MARYLAND-VIRGINIA AREA',
 'GREAT LAKES AREA',
 'HOUSTON AREA',
 'LOS ANGELES AREA',
 'MIDWEST AREA',
 'NEW ENGLAND AREA',
 'NEW YORK CITY AREA',
 'NORTH FLORIDA AREA',
 'NORTHWEST-ROCKY MOUNTAIN AREA',
 'OHIO AREA',
 'PHILADELPHIA AREA',
 'SOUTH FLORIDA AREA',
 'SOUTHWEST AREA',
 'TENNESSEE AREA']
lista_de_df = []
# Bucle sobre los símbolos
print(" --- Inicie el bucle sobre los símbolos --- ")
for i in areas:
    print("Procesando el símbolo: " + i)
    temp_df = pd.read_csv("/content/Bases/" + i + ".csv",index_col="Customer_ID")
    temp_df["area"] = i
    lista_de_df.append(temp_df)
    # Usando un salto de línea al final de esta cadena de caracteres por estética
    print(" --- Bucle completo sobre los símbolos --- \n")
    # Combinando en un solo DataFrame usando el concat
    print("Agregando los datos")
    agr_df = pd.concat(lista_de_df, axis=0)
    print(agr_df.shape)
    print("Cabeza del DataFrame agr_df: ")
agr_df.head()

Definición de los símbolos de las acciones
 --- Inicie el bucle sobre los símbolos --- 
Procesando el símbolo: ATLANTIC SOUTH AREA
 --- Bucle completo sobre los símbolos --- 

Agregando los datos
(592, 4)
Cabeza del DataFrame agr_df: 
Procesando el símbolo: CALIFORNIA NORTH AREA
 --- Bucle completo sobre los símbolos --- 

Agregando los datos
(1191, 4)
Cabeza del DataFrame agr_df: 
Procesando el símbolo: CENTRAL-SOUTH TEXAS AREA
 --- Bucle completo sobre los símbolos --- 

Agregando los datos
(1600, 4)
Cabeza del DataFrame agr_df: 
Procesando el símbolo: CHICAGO AREA
 --- Bucle completo sobre los símbolos --- 

Agregando los datos
(2096, 4)
Cabeza del DataFrame agr_df: 
Procesando el símbolo: DALLAS AREA
 --- Bucle completo sobre los símbolos --- 

Agregando los datos
(2614, 4)
Cabeza del DataFrame agr_df: 
Procesando el símbolo: DC-MARYLAND-VIRGINIA AREA
 --- Bucle completo sobre los símbolos --- 

Agregando los datos
(3254, 4)
Cabeza del DataFrame agr_df: 
Procesando el símbolo: GREA

mou_Mean  custcare_Mean new_cell                 area
Customer_ID                                                       
1069232       1776.50       3.666667        U  ATLANTIC SOUTH AREA
1043846          1.75       0.000000        U  ATLANTIC SOUTH AREA
1008539         93.00       0.000000        N  ATLANTIC SOUTH AREA
1068315        769.50       0.000000        U  ATLANTIC SOUTH AREA
1074220        586.00       1.666667        U  ATLANTIC SOUTH AREA

In [46]:
agr_df.sort_index(axis=0,inplace=True)
agr_df.head()

mou_Mean  custcare_Mean new_cell                           area
Customer_ID                                                                 
1000014           2.5       0.000000        Y       CENTRAL-SOUTH TEXAS AREA
1000016         270.5       1.000000        N                    DALLAS AREA
1000021        1978.0       0.333333        U                 TENNESSEE AREA
1000023        1771.5       0.000000        N  NORTHWEST-ROCKY MOUNTAIN AREA
1000029         608.0       0.000000        N               NEW ENGLAND AREA

In [108]:
agr_df

mou_Mean  custcare_Mean new_cell                           area
Customer_ID                                                                 
1000014           2.5       0.000000        Y       CENTRAL-SOUTH TEXAS AREA
1000016         270.5       1.000000        N                    DALLAS AREA
1000021        1978.0       0.333333        U                 TENNESSEE AREA
1000023        1771.5       0.000000        N  NORTHWEST-ROCKY MOUNTAIN AREA
1000029         608.0       0.000000        N               NEW ENGLAND AREA
...               ...            ...      ...                            ...
1099907        1893.0      17.333333        U               LOS ANGELES AREA
1099932           NaN       0.000000        U               LOS ANGELES AREA
1099963           NaN       0.000000        U                 SOUTHWEST AREA
1099981         492.0       0.333333        U                 SOUTHWEST AREA
1099992           NaN       0.000000        Y               LOS ANGELES AREA

[9997 rows x 4 columns]

La ejecución nos da como resultado una base de datos de 9997 registros correspondiente a la concatenación de todas las áreas. Se debe notar que nuestro conjunto de datos también es indexado por la columna `Customer_ID` y que la dimensión de los registros de está nueva base es menor a la de  `Creditos_dias.csv`.

 ## Cruce de información

El cruce de información es una operación esencial en el análisis de los datos. Usualmente, la posibilidad de realizar el cruce de dos o más fuentes de información es el primer paso para responder preguntas complejas acerca de los datos. Existen muchas formas de cruzar o combinar tablas de datos. El cruce de estos datos siempre tiene un propósito. Por tanto, se debe realizar un análisis de lo que se requiere antes de iniciar. Para esto, lo primero es conocer bien las fuentes de información que se van a relacionar.

### Uniones (Joins)
Usando la librería pandas se pueden realizar 4 tipos de uniones para cruzar información entre dos fuentes de datos, ver Figura abajo.
* INNER JOIN: este tipo de cruce incluye solo los resgistros que coincieden en las dos tablas
* LEFT JOIN:  en este tipo de cruce se incluyen todos los registros de la primera tabla aunque no aparezcan coincidencias en la segunda. Cuando no hay coincidencias en la segunda tabla los campos correspondientes a esta aparecerán como nulos.
* RIGHT JOIN: en este tipo de cruce se incluyen todos los registros de la segunda tabla aunque no aparezcan coincidencias en la primera. Cuando no hay coincidencias en la primer tabla los campos correspondientes a esta aparecerán como nulos.
* OUTER JOIN O FULL JOIN: combina todos los registros de ambas tablas aunque no existan coincidencias en la otra tabla.

![Los tipos de uniones que se pueden ejecutar en pandas](https://letsdobigdata.files.wordpress.com/2016/03/joins.png)

La combinación a usar depende del análisis que se requiera con los datos. Lo más común es realizar un cruce de tipo Inner Join, debido a que muestra los registros cuando aparecen relacionados en ambas tablas. Sin embargo, en
ocasiones se quiere saber que pasa con todos los registros de la primera tabla aunque no se tengan coincidencias. Alternativamente, quizá, lo menos usual es hacer combinaciones de tipo RIGHT o OUTER. Generalmente, estas son usadas cuando se quiere saber por qué no se tienen coincidencias, o cuáles registros no cumplen con la relación especificada.

Ya que desconocemos cuales son los registros perdidos en los 2 dataframes que hemos construido y es necesario para continuar nuestro preprocesamiento el unir la información.

In [47]:
df3=df.merge(agr_df,left_index=True,right_index=True,how="inner")
df3

creditcd  eqpdays  mou_Mean  custcare_Mean new_cell  \
Customer_ID                                                       
1015381            Y       49    728.00       0.000000        U   
1013369            Y      606     59.00       0.000000        U   
1025548            Y       81    959.25       0.333333        U   
1046337            Y      425    380.00       0.000000        Y   
1042815            Y      614    302.25       0.333333        Y   
...              ...      ...       ...            ...      ...   
1004659            Y      596    177.75       0.000000        N   
1006046            Y      622    247.00       0.000000        U   
1050240            Y       81    648.75       4.666667        U   
1081270            Y      119    368.25      23.000000        U   
1065953            N      394    135.75       0.000000        U   

                                      area  
Customer_ID                                 
1015381                       HOUSTON AREA  
1013369              CALIFORNIA NORTH AREA  
1025548                 NORTH FLORIDA AREA  
1046337                 NEW YORK CITY AREA  
1042815          DC-MARYLAND-VIRGINIA AREA  
...                                    ...  
1004659                          OHIO AREA  
1006046                 NEW YORK CITY AREA  
1050240          DC-MARYLAND-VIRGINIA AREA  
1081270                   LOS ANGELES AREA  
1065953      NORTHWEST-ROCKY MOUNTAIN AREA  

[1032 rows x 6 columns]

### Ejercicio 2  
¿Cuáles son los 3 registros que no se incluyen en df3 ?


In [48]:
sin_registros=(df.index.isin(df3.index))
df.iloc[~sin_registros,]

creditcd  eqpdays
Customer_ID                  
1021919            N      127
1061499            N      471
1089004            Y      289
1014402            Y      867
1014503            N      269
...              ...      ...
1088508            N      338
1094996            N      194
1006191            Y      953
1076628            N      285
1038465            N      472

[8968 rows x 2 columns]

Si deseamos realizar una filtración sobre algún valor relevante por ejemplo el área podremos revertir el proceso usando el operador `==`  que retorna `True` en el caso que dos objetos tengan el mismo valor

In [49]:
new_ingland=df3[df3["area"]=="NEW ENGLAND AREA"]
new_ingland.head()

creditcd  eqpdays  mou_Mean  custcare_Mean new_cell  \
Customer_ID                                                       
1095857            Y      234     59.75       0.000000        U   
1033914            Y      554    130.00       0.000000        U   
1093124            Y      290    688.75       7.333333        U   
1026602            Y      704    351.75       1.333333        U   
1085399            Y       10     87.00       0.000000        Y   

                         area  
Customer_ID                    
1095857      NEW ENGLAND AREA  
1033914      NEW ENGLAND AREA  
1093124      NEW ENGLAND AREA  
1026602      NEW ENGLAND AREA  
1085399      NEW ENGLAND AREA

El comando   `df3["area"]=="NEW ENGLAND AREA"`

In [50]:
df3["area"]=="NEW ENGLAND AREA"

Customer_ID
1015381    False
1013369    False
1025548    False
1046337    False
1042815    False
           ...  
1004659    False
1006046    False
1050240    False
1081270    False
1065953    False
Name: area, Length: 1032, dtype: bool

devuelve una serie boleana  con el mismo número de filas de la base`df3`, donde cada valor es `True` o `False` dependiendo si el valor de `area` de un registro  toma el valor de "NEW ENGLAND AREA"

### Agrupación por área geográfica

En el Analisis de la información se suelen usar funciones de resumen, las cuales plantean el agrupamiento de ciertas características por categorías. Para más información consultar [gropby().](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html).

Para nuestro desarrollo es de interés el revisar si el comportamiento de las variables se mantiene en cada una de las áreas geográficas.


In [51]:
df3.groupby("area")

Aquí, el objeto DataFrameGroupBy  contiene un objeto DataFrame para cada grupo (en este caso, un objeto DataFrame para cada símbolo). Específicamente, cada elemento del objeto es una tupla que contiene el identificador de grupo (en este caso el símbolo), y las filas correspondientes del DataFrame que tienen ese símbolo).

La librería `pandas`  permite iterar sobre el objeto groupby() para ver lo que hay dentro:

In [52]:
group_df = df3.groupby(["area"]) # Datos del grupo en df3 filtrados por el área
# Haciendo un bucle a través de los grupos
for item in group_df:
    print(type(item)) # Mostrando el tipo de artículo en df3
    print(item[0]) # área
    print(item[1].head()) # DataFrame

<class 'tuple'>
('ATLANTIC SOUTH AREA',)
            creditcd  eqpdays  mou_Mean  custcare_Mean new_cell  \
Customer_ID                                                       
1074053            Y      342   1264.75            0.0        U   
1073982            N      296    573.00            0.0        Y   
1025833            Y      371    386.25            0.0        U   
1030162            Y      634     47.00            0.0        Y   
1088785            Y      334     91.00            0.0        U   

                            area  
Customer_ID                       
1074053      ATLANTIC SOUTH AREA  
1073982      ATLANTIC SOUTH AREA  
1025833      ATLANTIC SOUTH AREA  
1030162      ATLANTIC SOUTH AREA  
1088785      ATLANTIC SOUTH AREA  
<class 'tuple'>
('CALIFORNIA NORTH AREA',)
            creditcd  eqpdays  mou_Mean  custcare_Mean new_cell  \
Customer_ID                                                       
1013369            Y      606     59.00       0.000000        U   


Usando el método  `groupby()` con el método `describe()` y apliquémoslo a cada símbolo para
analizar la distribución de las características la variable ----.

In [53]:
group_df = df3.groupby(["area"]) # Datos del grupo en df3 filtrados por el área
# Bucle a través de los grupos
for item in group_df:
    print("------Área: ", item[0])
    group_df1 = item[1]
    df_relevante = group_df[["totrev"]]
    print(df_relevante.describe())

------Área:  ('ATLANTIC SOUTH AREA',)


KeyError: "Columns not found: 'totrev'"

En `pandas`tenemos una manera más sencilla de mostrar estos resultados

In [ ]:
df3[["area","totrev"]].groupby("area").describe()

Estos resultados son iguales a los realizados con el ciclo `for`. Los objetos de tipo `DataFrameGroupBy` permiten una rápida salida de generación de estadísticas para grupos de interés.

### Ejercicio 3

Realice una agrupación por la variable creditcd_x y encuentre el valor de la desviación estándar

In [ ]:
# realice el ejercicio en este espacio

### Etiquetando valores de total de ingresos en categorías

Revisamos el comportamiento de `totrev` y encontramos una gran variabilidad, en ocasiones la agrupación de variables cuantitativas en diferentes categorías nos ayuda a tener una mejor visión del conjunto de datos.

Crearemos una nueva variable `Nivel_totrev`  que catalogue a la variable total de ingresos `totrev` en 2 posibles niveles, alto en el caso que `totrev`  sea mayor al cuantil 50 y baja en otro caso. cada valor de percentil debe ser calculado por cada `area` para asegurar que el +area geografíca sea evaluada adicionalmente.



In [ ]:
pencentiles50 = df3.groupby("area")["totrev"].quantile(0.5) # percentil 50
print(pencentiles50)

Como nos gustaría etiquetar los ingreos en alto y bajopara  cada área geográfica, haremos uso del método
`np.where()` en la biblioteca numpy. Este método toma una entrada y comprueba una condición lógica: si la
condición es verdadera, devolverá su segundo argumento, mientras que si la condición es falsa, devolverá su
tercer argumento.


In [ ]:
# Loop a través de los símbolos
areas = ['ATLANTIC SOUTH AREA',
 'CALIFORNIA NORTH AREA',
 'CENTRAL-SOUTH TEXAS AREA',
 'CHICAGO AREA',
 'DALLAS AREA',
 'DC-MARYLAND-VIRGINIA AREA',
 'GREAT LAKES AREA',
 'HOUSTON AREA',
 'LOS ANGELES AREA',
 'MIDWEST AREA',
 'NEW ENGLAND AREA',
 'NEW YORK CITY AREA',
 'NORTH FLORIDA AREA',
 'NORTHWEST-ROCKY MOUNTAIN AREA',
 'OHIO AREA',
 'PHILADELPHIA AREA',
 'SOUTH FLORIDA AREA',
 'SOUTHWEST AREA',
 'TENNESSEE AREA'] # registro de las áreas
lista_df = []
# ciclo sobre todos los símbolos
for i in areas:
    print("Etiqueta por área: " + i)
    temp_df = df3[df3["area"] == i] .copy()
    umbral_punto = pencentiles50.loc[i]
    temp_df["Nivel_totrev"] = np.where(temp_df["totrev"] < umbral_punto, "Bajo", "Alto")
    lista_df.append(temp_df)
df_con_etiquetas = pd.concat(lista_df)

In [ ]:
df_con_etiquetas.head()

In [ ]:
df_con_etiquetas.head()

Ahora podemos hacer una evaluaciónd de como es el comportamiento de los minutos gastados y el nivel de ingresos de los clientes

### ¿Existe relación alguna entre el total de minutos usados con la compañía y el ingreso total de los clientes?

Para explorar la relación entre el nivel de total de ingresos y el número promedio de llamadas, agrupemos por
Nivel_totrev y miremos ingreso total promedio por cada +area geográfica.



In [ ]:
df_con_etiquetas.groupby(['area','Nivel_totrev'])[['totmou']].mean()

### Ejercicio 4


Escriba el código para categorizar  el total de ingresos  baja, media y alta volatilidad, donde:

`
si totrev > (percentil 75 de totrev para el área dada):
Nivel_totrev = 'Alto'
o si VolStat > (percentil 25 de totrev para el área dada):
Nivel_totrev  = 'Medio'
de lo contrario:
Nivel_totrev  = 'Bajo'`

In [ ]:
# realice el ejercicio en este espacio

### Visualización de total de llamadas y total de ingresos

Ya hemos respondido satisfactoriamente a nuestra pregunta original. Sin embargo, no es necesario solamente
analizar los datos en formato tabular. Python contiene una funcionalidad que le permite analizar sus datos
visualmente también.

Usaremos la funcionalidad de pandas sobre la librería estándar de graficación de Python, `matplotlib`. Vamos
a importar la librería e instruir a Jupyter que muestre los gráficos en línea (es decir, mostrar los gráficos en
la pantalla del cuaderno para que podamos verlos mientras ejecutamos el código):



In [ ]:
modules = dir()
print(modules)

In [ ]:
import matplotlib.pyplot as plt
# Graficar en el cuaderno
%matplotlib inline

#### Histograma

In [ ]:
plt.style.use('ggplot')
plt.hist(df3["totrev"],bins=13,color="#237AAF")
plt.title('Histograma de variable totrev')
plt.xlabel("totrev");plt.ylabel("Frecuencia")
plt.show()

In [ ]:
print(plt.style.available)

#### Boxplot

In [ ]:
plt.style.use('ggplot')
plt.boxplot(df3["totrev"])
plt.title('Boxplot de variable totrev')
plt.show()

#### Gráficas a partir de `groupby`

In [ ]:
df4=df_con_etiquetas.groupby(['area','Nivel_totrev'])[['totmou']].mean()
df4.plot(kind='barh',figsize=(10,8))
plt.show()

#### Gráficas de `pandas`  

El modulo`.plot()` genera por defecto una gráfica de línea, existen varios gráficos para los `DataFrameGroup`

    - 'line' : Gráfica de línea
    - 'bar' :  Diagrama de barras vertical
    - 'barh' : Diagrama de barras horizontal
    - 'hist' : histograma
    - 'box' : boxplot
    - 'kde' : Kernel de densidad
    - 'area' : Gráfica de área
    - 'pie' : Diagrama de torta
    - 'scatter' : Diagrama de dispersión
    - 'hexbin' : Diagrama de hexágonos

In [ ]:
df3.columns

In [ ]:
df3["creditcd_x"].value_counts().plot(kind="pie")
plt.show()

Revisando la relación entre las dos variables `totrev` y `totmou`

In [ ]:
df3.plot(kind="scatter",x='totrev',y='totmou',c='blue')

### `Seaborn`

In [ ]:
import seaborn as sns
sns.boxplot(df3["totrev"],orient='v')
plt.title('Boxplot de variable totrev')
plt.show()

### Gráficas de parcela

In [ ]:
df3

In [ ]:
plt.figure(figsize=(116,4))
sns.pairplot(data=df3, hue='creditcd_x', vars=['totrev','totmou','eqpdays'])
plt.title("Parcelas")
plt.show()

In [ ]:
plt.figure(figsize=(6,4))
sns.boxplot(x='Nivel_totrev',y='totmou',data=df_con_etiquetas)
plt.title('Boxplot múltiple para Nivel de ingresos totales y \n total de minutos gastados')
plt.show()

In [ ]:
sns.catplot(y="totmou", x="Nivel_totrev", kind="box",    data=df_con_etiquetas, height=4, aspect=4, palette='Set2')
sns.catplot(y="totmou", x="Nivel_totrev", kind="violin", data=df_con_etiquetas, height=4, aspect=4, palette='Set2')

In [ ]:
ax = sns.catplot(y="totmou", x="Nivel_totrev", hue='creditcd_x', kind='point',data=df_con_etiquetas, aspect=2)
ax = sns.catplot(x="Nivel_totrev", kind='count',data=df_con_etiquetas, col='creditcd_x')

### Ejercicio 5

Realice un  boxplot para la variable totmou segmentado por las variable  credict_x y Nivel_totrev

In [ ]:
# realice el ejercicio en este espacio

## Resumen


1. Leer datos desde archivos.
2. Agregar y manipular datos .
3. Analizar estadísticas de resumen y reunir información.
4. Usar matplotlib para crear gráficos para hacer análisis visual, además de otras librerías.

Para más información

> [https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html)